_posts/2025-10-14-cqrs.adoc

```diff
@@ -0,0 +1,374 @@
+
+=== Replicating from PostgreSQL to QuestDB
+
+One of the advantages of using CQRS design pattern is that for aggregating and reading data we may use some special-purpose database which is able to optimize our query much better than general-purpose database.
+However, such database may not work with Debezium JDBC sink connector and if our database of choice is not MongoDB, it may look like we are out of luck.
+Fortunately not.
+We can take advantage of popularity of Kafka Connect platform.
+Due to Kafka popularity many database vendors provides their own sink connectors for Kafka Connect.
+
+In the second example, we replicate data from PostgreSQL into link:https://questdb.com/[QuestDB], a high-performance time-series database optimized for analytical queries.
+To keep out demo application simple and unified across all the examples, we don't use any QuestDB features or don't so provide any time-related charts, e.g. how the number of votes evolved during the time, but you can easily image that it would just a simple modification of the reader microservice to provide such plots.
+We cannot use JDBC sink connector for QuestDB as for ingesting the data it supports InfluxDB Line Protocol.
+However, QuestDB provides their own link:https://questdb.com/docs/third-party-tools/kafka/[sink connector] for Kafka Connect.
+Moreover it works well with Debezium format.
+
+QuestDB sink connector behaves little bit differently to Debezium sink connectors.
+It coverts into table columns all fields in the records.
+Thus, as the first step, we will modify our PostgreSQL source connector configuration little bit and flatten the record to contain only new record state.
+Conveniently Debezium provides link:https://debezium.io/documentation//reference/transformations/event-flattening.html[New record state SMT], which does exactly this - extracts only new state from the Debezium record.
+The configuration of the PostgreSQL source connector now looks like this:
+
+[source, json, options="nowrap"]
+----
+{
+    "name": "postgres-source",
+    "config": {
+        "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
+        "tasks.max": "1",
+        "database.hostname": "postgres",
+        "database.port": "5432",
+        "database.user": "postgres",
+        "database.password": "postgres",
+        "database.dbname" : "postgres",
+        "topic.prefix": "polls",
+        "table.include.list": "public.optionentity, public.pollentity, public.pollentity_optionentity, public.pollentity_voteentity, public.voteentity",
+        "transforms": "unwrap",
+        "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
+        "transforms.unwrap.add.fields": "table"
+    }
+}
+----
+
+In the SMT configuration we also add additional field - the name of the table which the record comes from.
+
+Another difference from the Debezium sink connectors is that single sink connector ingest data only to a single table.
```

Comment:
Are you sure this is the case? Reading the docs https://questdb.com/docs/third-party-tools/kafka/#configuration-options I'd say that by default there is topic to table mapping and everything can be routed into a single table but I might be wrong (also there is https://github.com/questdb/kafka-questdb-connector/blob/main/connector/src/main/java/io/questdb/kafka/Templating.java#L17).